### ISSUE TITLE
Install help under Win10 for Python beginners

### ISSUE THREAD

**Issue by Minga01** [vgavro/samsung-mdc#17]
Hi,
I'm completely new to Python, sorry. I was not able to install samsung-mdc.

What I tried so far:
- I installed "Git" and Python 3.12 from MS Store on my Windows 10 PC (22H2)
- I copied all files incl. source code into a created folder: c:\Samsung\python-samsung-mdc
- I opened "Git bash" box inside of this folder and tried your commands with it: 

> 
> [...]
> Minga@Windows-PC MINGW64 /c/Samsung
> $ echo alias samsung-mdc=\'python3 -m samsung_mdc\' >> ~/.bash_profile
> 
> Minga@Windows-PC MINGW64 /c/Samsung
> $ source ~/.bash_profile
> 
> Minga@Windows-PC MINGW64 /c/Samsung
> $ pip3 install python-samsung-mdc
> Defaulting to user installation because normal site-packages is not writeable
> WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002430BF395E0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/python-samsung-mdc/
> WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002430BFE5070>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/python-samsung-mdc/
> WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002430D23E4B0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/python-samsung-mdc/
> WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002430D23E690>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/python-samsung-mdc/
> WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000002430D23E8A0>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/python-samsung-mdc/
> ERROR: Could not find a version that satisfies the requirement python-samsung-mdc (from versions: none)
> ERROR: No matching distribution found for python-samsung-mdc
> [...]

I tried similar also with Windows' cmd.exe-box in same folder, with and without "--upgrade", also in sub-folder "samsung_mdc" (with underscore!?), double-clicked on "setup.py" => most end up with simliar errors like above!?

Btw. what I want to achieve with “samsung-mdc":
Just switch inputs of my Samsung QM55R from "HDMI1" to "DisplayPort1", and back, via network-cable, via command-line, via Elgato Stream Deck. 
(SNMP is enabled, I can ping the monitor, and I can switch via Samsung MDC-Software, but this is not as nice as via Stream Deck.)

e.g. something like this:
samsung-mdc 1@192.168.137.99:1515 input_source DisplayPort

(btw. "DisplayPort" is written on the OSD at the monitor, but this input is called "DisplayPort1" inside Samsung MDC software. I will try both once samsung-mdc is working. ;)

Maybe you have an idea?

**Comment by vgavro** [comment]
it seems like your pip is broken for whatever reason.

**Comment by Minga01** [comment]
Thanks for your answer, I will check next week for any set proxy like said on your hyperlink, but I don't think so, because a proxy was never added before. The PC is offline, only inside a test LAN. Do I need to be connected to the internet for the install process? 

I don't think “numpy” or “anaconda” will help me as a noob, I would rather try without further levels of complication. ;)

Or is there another way to install it under Windows? pip or pip3? Any typo or folder issue on my side?
Is your progam able to run without "installation", e.g. like "python -m samsung-mdc" ? (idea adapted from https://github.com/kdschlosser/samsungctl)

I can no longer find the source on the Internet, but somewhere it said that Samsung's SNMP implementation was solved “impractically”, so I find the control via this “MDC protocol” promising. At least MDC works.

Have a nice weekend!

**Comment by vgavro** [comment]
pip is "python package manager". python-samsung-mdc (this repository) distributed as "python package".
Basically what's happening here, is package manager on installing package returns to you connection errors while trying to connect to packages repository to install package by it's name. You understand that your pc is offline. So either get it online, either google ways to install python package (and it's dependencies!) to offline machine. Out of scope this readme, maybe I'll add note that "you should be online" for pip3 install command.

**Comment by Minga01** [comment]
But i got other strange issue now, e.g. if I ask for the status:
> $ samsung_mdc 1@192.168.137.99:1515 status
I get
> 1@192.168.137.99:1515 MDCTLSRequired: b'MDCSTART<<TLS>>'

Btw. in my folder
c:\Users\xyz\Appdata\Local\Microsoft\WindowsApps\
I have now 9x "p*.exe", all 0 KB, all from 08.10.2024:
pip.exe, pip3.12.exe, pip3.exe, python.exe, python3.12.exe, python3.exe, pythonw.exe, pythonw3.12.exe, pythonw3.exe
 
I would prefer to use Windows' MS-DOS-style command-window instead of Git Bash, even if I have to add complete long folder-paths. If this is possible with your program, please add the syntax in your description.

**Comment by vgavro** [comment]
> But I got other strange issue now, e.g. if I ask for the status:
> 
> > $ samsung_mdc 1@192.168.137.99:1515 status
> 
> I get
> 
> > 1@192.168.137.99:1515 MDCTLSRequired: b'MDCSTART<>'

Your device requires "PIN" to create secure connection. You can see what PIN required (or turn it off) on device GUI.
Use --pin to provide pin from command line.

**Comment by Minga01** [comment]
After disabling „Security“ in the monitor-GUI it works, thanks a lot! :)

Btw. I had to switch in the monitor-GUI from my set “prefer DisplayPort as input source” back to “None”, otherwise my switch to HDMI doesn’t stay long, only a few seconds, and then switches back to DisplayPort.

> $ samsung-mdc 1@192.168.137.99:1515 input_source HDMI1
> 1@192.168.137.99:1515 <INPUT_SOURCE_STATE.HDMI1:33>
> 

Btw. the original Samsung-MDC-software doesn’t care what is set in the monitor-GUI.
But this Samsung-MDC-software needs security via SNMP to be enabled, so I put it back ON, to be able to work also with MDC in parallel. But I was not able to send the PIN (default 0000):

> $ samsung-mdc -p 0000 1@192.168.137.99:1515 input_source Display_Port_1
> 1@192.168.137.99:1515 SSLError: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1000)
> 

Maybe a stupid question from me as a Python newbie: This “install” at the beginning is not a permanent installation in the classic sense, but rather a “run”, i.e. after a reboot I have to “install” your “samsung-mdc” again?

And is there perhaps an even easier way, without installation, to call “commands.py” or “__ main __.py” or similar, directly in order to change the monitor-input-source?

FYE: If you do something like this nonsense out of desperation, you get the following result:
(in cmd.exe-window:)

> C:\Samsung>C:\Users\xyz\AppData\Local\Microsoft\WindowsApps\python3.12.exe "C:/Samsung/python-samsung-mdc/samsung_mdc"
> Traceback (most recent call last):
>   File "<frozen runpy>", line 198, in _run_module_as_main
>   File "<frozen runpy>", line 88, in _run_code
>   File "C:\Samsung\python-samsung-mdc\samsung_mdc\__main__.py", line 2, in <module>
>     from .cli import cli
> ImportError: attempted relative import with no known parent package
>

**Comment by vgavro** [comment]
> After disabling „Security“ in the monitor-GUI it works, thanks a lot! :)
> 
> Btw. I had to switch in the monitor-GUI from my set “prefer DisplayPort as input source” back to “None”, otherwise my switch to HDMI doesn’t stay long, only a few seconds, and then switches back to DisplayPort.
> 
> > $ samsung-mdc 1@192.168.137.99:1515 input_source HDMI1
> > 1@192.168.137.99:1515 <INPUT_SOURCE_STATE.HDMI1:33>
> 
> Btw. the original Samsung-MDC-software doesn’t care what is set in the monitor-GUI. But this Samsung-MDC-software needs security via SNMP to be enabled, so I put it back ON, to be able to work also with MDC in parallel. But I was not able to send the PIN (default 0000):
> 
> > $ samsung-mdc -p 0000 1@192.168.137.99:1515 input_source Display_Port_1
> > 1@192.168.137.99:1515 SSLError: [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:1000)
> 
 Thanks for reporting this, i'm not actually sure what error will be displayed on wrong pin, maybe try to set some pin and check it again?
The problem with ssl/pin authentication is that it's not documented, but it was tested on devices I had at that time to work same as GUI reference application... Not to say they may change this in future (already) for new devices, and reference application isn't publicly available.
I'll try to test it when I'll be near some Samsung device next time, but plz test with non-default pin before that, and report back results, thank you!

**Comment by vgavro** [comment]
> Maybe a stupid question from me as a Python newbie: This “install” at the beginning is not a permanent installation in the classic sense, but rather a “run”, i.e. after a reboot I have to “install” your “samsung-mdc” again?
> 
> And is there perhaps an even easier way, without installation, to call “commands.py” or “__ main __.py” or similar, directly in order to change the monitor-input-source?
> 
> FYE: If you do something like this nonsense out of desperation, you get the following result: (in cmd.exe-window:)
> 
> > C:\Samsung>C:\Users\xyz\AppData\Local\Microsoft\WindowsApps\python3.12.exe "C:/Samsung/python-samsung-mdc/samsung_mdc"
> > Traceback (most recent call last):
> > File "", line 198, in _run_module_as_main
> > File "", line 88, in _run_code
> > File "C:\Samsung\python-samsung-mdc\samsung_mdc__main__.py", line 2, in 
> > from .cli import cli
> > ImportError: attempted relative import with no known parent package
 Package is installed one-time, you can see it available in your (Python-Path)/Scripts directory, also.you can just add it to your PATH env variable, if you can't see "Samsung-mdc" there - ensure that "pip" you're using is actually using same Python-Path maybe?
